storage/store.go

```diff
@@ -489,23 +493,97 @@ func (s *Store) Start(stopper *util.Stopper) error {
 // startGossip runs an infinite loop in a goroutine which regularly checks
 // whether the store has a first range replica and asks that range to gossip
 // cluster ID and first range metadata accordingly.
-func (s *Store) startGossip() {
+func (s *Store) startGossip() error {
+	// Take care of gossip for config descriptors. It does not
+	// expire, so this is a one-time action in a working cluster - if values change, ranges
+	// will update gossip autonomously.
+	// However, the lease holder, who is normally in charge of that
+	// might crash precisely when trying to gossip. To account for
+	// this rare scenario, we activate all ranges that hold config
+	// maps periodically.
+	configFn := func() error {
+		for _, cd := range configDescriptors {
+			// If we ever allow configs to span multiple ranges, we'll need to
+			// get them all. Below we assume that one exists containing this
+			// whole chunk.
+			rng := s.LookupRange(cd.keyPrefix, cd.keyPrefix.Next())
+			if rng == nil {
+				log.Warningf("no range")
+				// This store has no range with this configuration.
+				continue
+			}
+			// Wake up the replica. If it acquires a fresh lease, it will
+			// gossip. If an unexpected error occurs (i.e. nobody else seems to
+			// have an active lease but we still failed to obtain it), return
+			// that error. If we ignored it we would run the risk of running a
+			// cluster without configs gossiped.
+			if _, err := rng.getLeaseForGossip(); err != nil {
+				return err
+			}
+			// Since leader leases are persisted, and a bootstrapping range
+			// gets a leader lease, we can't actually trust that the above
+			// automatically gossips the configs. The initial start of a fresh
+			// cluster after bootstrap, if fast enough, will actually be happy
+			// with its previous lease and not automatically re-gossip. It has
+			// an empty gossip though. TestStatusGossipJson fails without this.
+			// If a leader crashes right after applying a leader lease and
+			// gossiping, reboots and still holds the lease, the same effect
+			// occurs.
+			rng.maybeGossipConfigs(func(_ proto.Key) bool {
```

Comment:
That's possible but needs some refactoring. I had toyed with something like this but since we gossip configs in InternalLeaderLease, and getLeaseForGossips calls redirectOnOrAcquireLeaderLease which has a lock (and is already running most of the time you end up in InternalLeaderLease), you end up deadlocking. You'd need to find a new place for config gossip on lease acquiry to live unless you want to special case the gossiping from InternalLeaderLease or remove the locking in redirectOnOrAcquireLeaderLease.